AutoDJProcessor: Calculate and expose the number of tracks and accurate total duration in the Auto DJ queue

cr7pt0gr4ph7 opened the pull request:
The calculated total duration take into account the transition modes, transition time etc. (in fact, it reuses the code that performs the actual transition during queue playing).

The `queue_duration` ~~`time_remaining`~~ and `queue_tracks` ~~`tracks_remaining`~~ are exposed as `ControlObject`s so they can be consumed by skins.

cr7pt0gr4ph7 committed:
AutoDJFeature: Fix spelling of 'findOrCreateAutoDjPlaylistId'

cr7pt0gr4ph7 committed:
AutoDJProcessor: Feature: Add control objects for Auto DJ queue status

The time_remaining and tracks_remaining controls in the [AutoDJ]
can be used to display the information about the Auto DJ queue
in custom skins.

The current calculation method for the time_remaining control
only considers the total duration of the tracks by themselves,
and does not take the transition times between tracks into account.

cr7pt0gr4ph7 committed:
AutoDJFeature: Feature: Append the duration & track count to the AutoDJ tree item

cr7pt0gr4ph7 committed:
AutoDJProcessor: Add TrackOrDeckAttributes

This is a first step in generalizing the transition logic in
AutoDJProcessor to enable it to also be applied to tracks in
the queue instead of just the current decks.

cr7pt0gr4ph7 committed:
AutoDJProcessor: Add FadeableTrackOrDeckAttributes

This is another step in generalizing the transition logic
in AutoDJProcessor to enable calculating transition times
for the whole queue.

cr7pt0gr4ph7 committed:
AutoDJProcessor: Add calculateTransitionImpl

This is the final step in generalizing the core transition timing
logic in AutoDJProcessor, so that it can be applied to a whole
playlist instead of just the current decks.

cr7pt0gr4ph7 committed:
AutoDJProcessor: Feature: Accurately calculate the remaining play time

cr7pt0gr4ph7 committed:
AutoDJProcessor: Trigger recalculation of the remaining play time when relevant

The following data points are used as inputs for the "remaining time"
calculation, and should therefore trigger a recalculation:

* The list of tracks in the Auto DJ playlist (both
  the set of tracks and their order are important).

  We subscribe to PlaylistTableModel::tracksChanged
  to receive the relevant notifications.

* The AutoDJ transition mode and transition time settings.
  We will be notified via ::setTransitionMode
  and ::setTransitionTime, respectively.

* The Intro, Outro & N60dBSound cues of all tracks
  that are contained in the Auto DJ queue.

  We subscribe to TrackCollection::tracksChanged
  and TrackCollection::multipleTracksChanged to
  receive notifications about possible changes.

  As of now, we do not filter the notifications,
  and simply trigger a recalculation when ANY
  track has changed.

cr7pt0gr4ph7 committed:
AutoDJProcessor: Readability: Append "Second" suffix to variables

cr7pt0gr4ph7 committed:
AutoDJProcessor: Readability: Shorten variable names

cr7pt0gr4ph7 committed:
AutoDJProcessor: Expose getter methods in addition to ControlObjects for the remaining time & tracks

cr7pt0gr4ph7 committed:
DlgAutoDJ: Feature: Display the total duration of the Auto DJ queue (accounting for transition times)

cr7pt0gr4ph7 commented:
@ninomp Maybe you can have a look at reviewing this PR

daschuer commented:
Oh sorry, a conflict has developed.

daschuer commented on src/library/autodj/autodjprocessor.cpp:
```diff
@@ -16,21 +18,95 @@ const char* kTransitionPreferenceName = "Transition";
 constexpr bool sDebug = false;
 } // anonymous namespace
 
+TrackOrDeckAttributes::~TrackOrDeckAttributes() {
```
This is probably redundant

daschuer replied:
Normally we use one file per class

daschuer reviewed (changes requested):
This is a good start. I am afraid the calculation is not yet correct in any case. Did you consider to reuse the original transition calculation code? 

This PR includes a lot of useful refactorings. 
I have interest to merge them first in a separate PR. Can you prepare that? 

What is the use case of the new duration controls?

cr7pt0gr4ph7 commented:
Thanks for having a look at reviewing this :+1: 

> Oh sorry, a conflict has developed.

Thats one reason why I want it merged into upstream, at least at one point in the future :sweat_smile: I will have a go at resolving the conflicts as soon as I find time again.

> This is a good start. I am afraid the calculation is not yet correct in any case. Did you consider to reuse the original transition calculation code?

Uh, I am not sure what you mean. Just to make sure that I understand you correctly:

I actually use the original transition calculation code, but had to refactor it so that I could apply it not only to tracks already loaded into a deck, but to playlist tracks as well.

These changes are isolated into the commits 5c6ce7a ([review link](https://github.com/mixxxdj/mixxx/pull/13183/commits/5c6ce7a41d615f28782ae17b8341ca6c412ca9fd)), b646631 ([review link](https://github.com/mixxxdj/mixxx/pull/13183/commits/b6466311beddf45dfcf62fe47989c0414aa616bf)) and 3eba5f2a ([review link](https://github.com/mixxxdj/mixxx/pull/13183/commits/3eba5f2ae4f9da382d3826b761a3c9568d571007)), which I tried to keep small and self contained.

I have also tested the feature with different transition modes, transition times etc., and haven't noticed any problems.

Is there a specific part of the code that worries you?

> This PR includes a lot of useful refactorings. I have interest to merge them first in a separate PR. Can you prepare that?

Sure. I actually already structured the commits in a way that would make this possible. Which changes would you like to get in a separate PR?

> What is the use case of the new duration controls?

To display them on DJ controllers and/or in skins and/or in a status bar (see also #13176 for an example of how it could be done).

cr7pt0gr4ph7 committed:
AutoDJConstants: Extract common constants into AutoDJConstants.

cr7pt0gr4ph7 committed:
TrackOrDeckAttributes: Move TrackOrDeckAttributes and subclasses into separate files.

cr7pt0gr4ph7 commented:
@daschuer I've resolved the merge conflicts and moved `TrackOrDeckAttributes` and its subclasses into separate files.

As mentioned in the previous comment, this reuses the original transition code for both the actual transitions, and the ahead-of-time playlist length calculations (see [autoprocessor.cpp:L1239-L1503](https://github.com/mixxxdj/mixxx/pull/13183/files#diff-0468b47d7671b5286de23e574e3b1062d8fe118040f4cab3921ce60265eba4a6L1239-L1503)). I've only renamed variables like `pFromDeck -> fromTrack`, and added `TrackOrDeckAttributes` and `TrackAttributes` to generalize code that was previously using `DeckAttributes` where possible. In every other regard, the transition code _should_ behave 1:1 as before.

There _is_ one (small) inconsistency, which is currently by design:
- The `Remaining Queue Playtime` calculation is done based solely on the contents of the Auto DJ queue and the configured transition time times between its tracks. In particular, it does **not** take the currently playing track(s)[^1] or the transition time from the currently playing track to the next track in the queue into account.

[^1]: Sidenote: Determining the "current track" from which we will transition to our next track in the queue is only possible when Auto DJ is either already active, or could be activated. It is not possible when e.g. multiple tracks are currently playing.

After thinking about it, there are two different usecases, only one of which is fully solved here - and it might make sense to rename our control objects to allow implementing the other one in the future:

| | Use Case 1 | Use Case 2 |
|-- | -- | -- |
| What? | Length/Duration of Auto DJ queue | Remaining playtime in Auto DJ mode |
| When is it relevant? (Use case) | **Preparing an Auto DJ queue** while something else/unrelated is playing: _"How long will this list of tracks be playing **after** I start it?"_ | **While playing using Auto DJ**: _"How much time do I have until music stops?"_ |
| When does this number change? | Discrete: Whenever the track list changes, i.e. when loading the next track | Continuously: All the time (while playing) | 
| Solved by this PR? | Yes | Partially (the user has to manually add the remaining track time to the remaining queue time in their head) |

cr7pt0gr4ph7 committed:
AutoDJProcessor: Rename tracks_remaining/time_remaining COs to queue_tracks/queue_duration.

cr7pt0gr4ph7 committed:
AutoDJProcessor: Fix a few spelling and grammar errors in comments.

cr7pt0gr4ph7 committed:
AutoDJProcessor: Fix: Correctly calculate the time adjustment due to transitions

cr7pt0gr4ph7 commented:
As it turns out, the transition calculation code seems to work fine, but I had messed up the small piece of new code that calculates how a given transition between two tracks affects their total playtime... fixed now.

cr7pt0gr4ph7 commented:
Can I do anything to move this PR forward/get it merged? It's been one month....
Implementation is complete, I'm running this in my production setup, and haven't encountered any problems so far.

_**PS:** If you want to quickly try it out without having to build it yourself, you can use the `io.github.cr7pt0gr4ph7.Mixxx` Flatpak published at [`cr7pt0gr4ph7/mixxx-flatpak`](https://github.com/cr7pt0gr4ph7/mixxx-flatpak) that includes this and a few other changes. Parallel installation is possible, i.e. any existing official or Flatpak installation of Mixxx will not be affected:_
```
sudo flatpak remote-add --if-not-exists mixxx-flatpak https://cr7pt0gr4ph7.github.io/mixxx-flatpak/default.flatpakrepo
sudo flatpak install io.github.cr7pt0gr4ph7.Mixxx
```

cr7pt0gr4ph7 commented:
@ninomp Would you be willing to give this PR another look, to see if it can be merged? Thanks! :)

daschuer commented:
This is a big PR with Lines changed: 882 additions & 410 deletions. 
It is hard to review and test. 
There is also a comment that the calculation is not yet correct. I don't know what the current state is. 
If you want to take over, a first step is to move the prerequisite refactoring into a separate PRs. 
Than we can probably merge that finally in steps.

cr7pt0gr4ph7 commented:
@daschuer

> This is a big PR with Lines changed: 882 additions & 410 deletions. It is hard to review and test. There is also a comment that the calculation is not yet correct. I don't know what the current state is. If you want to take over, a first step is to move the prerequisite refactoring into a separate PRs. Than we can probably merge that finally in steps.

Thanks, I'll try to get a few, smaller PRs going.

cr7pt0gr4ph7 commented:
Closing as the changes will land as separate PRs. See #16865 for follow-up.
